utils.go

```diff
@@ -122,52 +124,53 @@ const defaultSymFile = "/proc/kallsyms"
 // for older syscall functions to run on newer kernels
 func getSyscallName(name string, symFile string) (string, error) {
 	// Get kernel symbols
-	syms, err := os.ReadFile(symFile)
+	syms, err := os.Open(symFile)
 	if err != nil {
 		return "", err
 	}
-	return getSyscallFnNameWithKallsyms(name, string(syms))
+	defer syms.Close()
+
+	return getSyscallFnNameWithKallsyms(name, syms, "")
 }
 
-func getSyscallFnNameWithKallsyms(name string, kallsymsContent string) (string, error) {
-	var arch string
-	switch runtime.GOARCH {
-	case "386":
-		arch = "ia32"
-	case "arm64":
-		arch = "arm64"
-	default:
-		arch = "x64"
+func getSyscallFnNameWithKallsyms(name string, kallsymsContent io.Reader, arch string) (string, error) {
+	if arch == "" {
+		switch runtime.GOARCH {
+		case "386":
+			arch = "ia32"
+		case "arm64":
+			arch = "arm64"
+		default:
+			arch = "x64"
+		}
 	}
 
 	// We should search for new syscall function like "__x64__sys_open"
 	// Note the start of word boundary. Should return exactly one string
-	regexStr := `(\b__` + arch + `_[Ss]y[sS]_` + name + `\b)`
-	fnRegex := regexp.MustCompile(regexStr)
+	newSyscall := regexp.MustCompile(`\b__` + arch + `_[Ss]y[sS]_` + name + `\b`)
+	// If nothing found, search for old syscall function to be sure
+	oldSyscall := regexp.MustCompile(`\b[Ss]y[sS]_` + name + `\b`)
+	// check for '__' prefixed functions, like '__sys_open'
+	prefixed := regexp.MustCompile(`\b__[Ss]y[sS]_` + name + `\b`)
 
-	match := fnRegex.FindAllString(kallsymsContent, -1)
-	if len(match) > 0 {
-		return match[0], nil
-	}
+	scanner := bufio.NewScanner(kallsymsContent)
+	scanner.Split(bufio.ScanLines)
 
-	// If nothing found, search for old syscall function to be sure
-	regexStr = `(\b[Ss]y[sS]_` + name + `\b)`
-	fnRegex = regexp.MustCompile(regexStr)
-	match = fnRegex.FindAllString(kallsymsContent, -1)
-	// If we get something like 'sys_open' or 'SyS_open', return
-	// either (they have same addr) else, just return original string
-	if len(match) > 0 {
-		return match[0], nil
-	}
+	for scanner.Scan() {
+		line := scanner.Text()
 
-	// check for '__' prefixed functions, like '__sys_open'
-	regexStr = `(\b__[Ss]y[sS]_` + name + `\b)`
-	fnRegex = regexp.MustCompile(regexStr)
-	match = fnRegex.FindAllString(kallsymsContent, -1)
-	// If we get something like '__sys_open' or '__SyS_open', return
-	// either (they have same addr) else, just return original string
-	if len(match) > 0 {
-		return match[0], nil
+		if !strings.Contains(line, name) {
+			continue
+		}
+
+		for _, pattern := range []*regexp.Regexp{newSyscall, oldSyscall, prefixed} {
```

Comment:
Do you think the compiler is smart enough to avoid re-creating this slice each time? Maybe hoist it outside the for loop to be safe?

Reply:
One difference in behavior, is this finds the first line that matches any of the regexes. The previous behavior would search the entire file for a new match, then an old match, then prefixed.

Reply:
I would hope it's able to hoist it but let's not take the risk.

For the behavior change I think it doesn't really matter since we do not expect multiple conflicting patterns to occur at the same time, I also don't see a way to keep the old behavior without reading the file multiple times which would be not great

Reply:
Hmm, my ubuntu 22.04 machine has two of them:
```
0000000000000000 T __sys_connect
0000000000000000 T __arm64_sys_connect
```

Reply:
One idea is to read through the entire file (with the scanner) and:
1. If we get the arch-specific one, return immediately since that is the priority.
2. otherwise collect everything that matches and then apply the priority to that subset.

This should still be much better from a memory standpoint, and is ~one pass, since the subset should be really small.

Reply:
Ah this was not the case on my end but if my assumption is wrong then clearly we need a better way. I will implement something similar to what you described here yes (like get the result for each regex, and do the priority at the end)